pallet/ethy/src/lib.rs

```diff
@@ -456,112 +695,177 @@ decl_module! {
+						ParamType::Address,
+						ethabi::ParamType::Bytes,
+						ParamType::Uint(64),
+					],
+					event.as_slice(),
+				)
+				.map_err(|_| Error::<T>::InvalidClaim)?
+				.as_slice()
+			{
 				let event_id: EventClaimId = (*event_id).saturated_into();
-				ensure!(!PendingEventClaims::contains_key(event_id), Error::<T>::EventReplayPending); // NOTE(surangap): prune PendingEventClaims also?
-				if !Self::processed_message_ids().is_empty() {
-					ensure!( event_id > Self::processed_message_ids()[0] &&
-						Self::processed_message_ids().binary_search(&event_id).is_err() , Error::<T>::EventReplayProcessed);
+				ensure!(
+					!PendingEventClaims::<T>::contains_key(event_id),
+					Error::<T>::EventReplayPending
+				); // NOTE(surangap): prune PendingEventClaims also?
+				if !ProcessedMessageIds::<T>::get().is_empty() {
+					ensure!(
+						event_id > ProcessedMessageIds::<T>::get()[0] &&
+							ProcessedMessageIds::<T>::get().binary_search(&event_id).is_err(),
+						Error::<T>::EventReplayProcessed
+					);
 				}
+				let data = BoundedVec::try_from(data.as_slice().to_vec())
+					.map_err(|_| Error::<T>::InvalidClaim)?;
 				let event_claim = EventClaim {
 					tx_hash,
 					source: *source,
 					destination: *destination,
 					data: data.clone(),
 				};
 
-				PendingEventClaims::insert(event_id, &event_claim);
-				PendingClaimStatus::insert(event_id, EventClaimStatus::Pending);
+				PendingEventClaims::<T>::insert(event_id, &event_claim);
+				PendingClaimStatus::<T>::insert(event_id, EventClaimStatus::Pending);
 
 				// TODO: there should be some limit per block
-				let process_at: T::BlockNumber = <frame_system::Pallet<T>>::block_number() + Self::challenge_period();
-				<MessagesValidAt<T>>::append(process_at, event_id);
-
-				Self::deposit_event(Event::<T>::EventSubmit(event_id, event_claim, process_at));
+				let process_at: T::BlockNumber =
+					<frame_system::Pallet<T>>::block_number() + ChallengePeriod::<T>::get();
+				MessagesValidAt::<T>::mutate(process_at.clone(), |v| {
+					let mut message_ids = v.clone().into_inner();
+					message_ids.push(event_id);
+					let message_ids_bounded = WeakBoundedVec::force_from(
+						message_ids,
+						Some(
+							"Warning: There are more MessagesValidAt than expected. \
+								A runtime configuration adjustment may be needed.",
+						),
+					);
+					*v = message_ids_bounded;
+				});
+
+				Self::deposit_event(Event::<T>::EventSubmit {
+					event_claim_id: event_id,
+					event_claim,
+					process_at,
+				});
 			}
+			Ok(())
 		}
 
-		#[weight = DbWeight::get().writes(1) + DbWeight::get().reads(2)]
 		/// Submit a challenge for an event
 		/// Challenged events won't be processed until verified by validators
 		/// An event can only be challenged once
-		pub fn submit_challenge(origin, event_claim_id: EventClaimId) -> DispatchResult {
+		#[pallet::weight(DbWeight::get().writes(1) + DbWeight::get().reads(2))]
+		pub fn submit_challenge(
```

Comment:
should be transactional
```suggestion
                #[transactional]
		pub fn submit_challenge(
```